File: app/src/main/java/com/ducktapedapps/updoot/binding/SubredditBindingAdapters.java
package com.ducktapedapps.updoot.binding;

import android.graphics.Color;
import android.text.format.DateUtils;
import android.widget.TextView;

import androidx.core.content.ContextCompat;
import androidx.databinding.BindingAdapter;

import com.ducktapedapps.updoot.R;
import com.ducktapedapps.updoot.model.LinkData;
import com.ducktapedapps.updoot.utils.MarkdownUtils;

public class SubredditBindingAdapters {
    @BindingAdapter("metadata")
    public static void setMetadata(TextView textView, LinkData data) {
        StringBuilder metadata = new StringBuilder().append(data.getSubredditName());
        metadata.append(" \u2022 ");
        if (data.getCommentsCount() > 999) {
            metadata
                    .append(data.getCommentsCount() / 1000)
                    .append("K replies");
        } else {
            metadata.append(data.getCommentsCount())
                    .append(" replies");
        }
        metadata.append(" \u2022 ");

        metadata.append(DateUtils.getRelativeTimeSpanString(data.getCreated() * 1000, System.currentTimeMillis(), DateUtils.MINUTE_IN_MILLIS))
                .append(" ");
        textView.setText(metadata.toString());
    }

    @BindingAdapter({"submissionVotes", "voteColor"})
    public static void setVotes(TextView textView, int votes, Boolean likes) {
        String upVotes;
        if (votes > 999) {
            upVotes = votes / 1000 + "K";
        } else {
            upVotes = String.valueOf(votes);
        }

        if (likes == null) {
            textView.setTextColor(Color.WHITE);
        } else if (likes) {
            textView.setTextColor(ContextCompat.getColor(textView.getContext(), R.color.upVoteColor));
        } else {
            textView.setTextColor(ContextCompat.getColor(textView.getContext(), R.color.downVoteColor));
        }

        textView.setText(upVotes);
    }

    @BindingAdapter({"toggleSelfTextVisibility", "selfText"})
    public static void toggleSelfTextVisibility(TextView selfTextView, boolean isSelfTextExpanded, String htmlSelfText) {
        if (isSelfTextExpanded) {
            MarkdownUtils.decodeAndSet(htmlSelfText, selfTextView);
        } else {
            selfTextView.setText("");
        }
    }
}

File: app/src/main/java/com/ducktapedapps/updoot/binding/SubredditBindingAdapters.kt
package com.ducktapedapps.updoot.binding

import android.graphics.Color
import android.text.format.DateUtils
import android.widget.TextView

import androidx.core.content.ContextCompat
import androidx.databinding.BindingAdapter

import com.ducktapedapps.updoot.R
import com.ducktapedapps.updoot.model.LinkData
import com.ducktapedapps.updoot.utils.MarkdownUtils

object SubredditBindingAdapters {
    @JvmStatic
    @BindingAdapter("metadata")
    fun setMetadata(textView: TextView, data: LinkData) {
        val metadata = StringBuilder().append(data.subredditName)
        metadata.append(" \u2022 ")
        if (data.commentsCount > 999) {
            metadata
                    .append(data.commentsCount / 1000)
                    .append("K replies")
        } else {
            metadata.append(data.commentsCount)
                    .append(" replies")
        }
        metadata.append(" \u2022 ")

        metadata.append(DateUtils.getRelativeTimeSpanString(data.created * 1000, System.currentTimeMillis(), DateUtils.MINUTE_IN_MILLIS))
                .append(" ")
        textView.text = metadata.toString()
    }

    @JvmStatic
    @BindingAdapter("submissionVotes", "voteColor")
    fun setVotes(textView: TextView, votes: Int, likes: Boolean?) {
        val upVotes: String = if (votes > 999) {
            (votes / 1000).toString() + "K"
        } else {
            votes.toString()
        }

        when {
            likes == null -> textView.setTextColor(Color.WHITE)
            likes -> textView.setTextColor(ContextCompat.getColor(textView.context, R.color.upVoteColor))
            else -> textView.setTextColor(ContextCompat.getColor(textView.context, R.color.downVoteColor))
        }

        textView.text = upVotes
    }

    @JvmStatic
    @BindingAdapter("toggleSelfTextVisibility", "selfText")
    fun toggleSelfTextVisibility(selfTextView: TextView, isSelfTextExpanded: Boolean, htmlSelfText: String?) {
        if (isSelfTextExpanded) {
            MarkdownUtils.decodeAndSet(htmlSelfText ?: "", selfTextView)
        } else {
            selfTextView.text = ""
        }
    }
}
